packages/agent/test/agent_spec.ts

```diff
@@ -0,0 +1,181 @@
+    context(testCase.name, function() {
+      beforeEach(function() {
+        this.oldEnv = Object.assign({}, process.env)
+        process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
+
+        if (testCase.proxyUrl) {
+          process.env.HTTP_PROXY = process.env.HTTPS_PROXY = testCase.proxyUrl
+          process.env.NO_PROXY = ''
+        }
+
+        this.agent = new CombinedAgent()
+
+        this.request = request.defaults({
+          proxy: null,
+          strictSSL: false,
+          agent: this.agent
+        })
+
+        if (testCase.proxyUrl) {
+          let options: any = {
+            keepRequests: true,
+            https: false,
+            auth: false
+          }
+
+          if (testCase.httpsProxy) {
+            options.https = this.https
+          }
+
+          if (testCase.proxyAuth) {
+            options.auth = {
+              username: 'foo',
+              password: 'bar'
+            }
+          }
+
+          this.debugProxy = new DebuggingProxy(options)
+          return this.debugProxy.start(PROXY_PORT)
+        }
+      })
+
+      afterEach(function() {
+        process.env = this.oldEnv
+
+        if (testCase.proxyUrl) {
+          this.debugProxy.stop()
+        }
+      })
+
+      it('HTTP pages can be loaded', function() {
+        return this.request({
+          url: `http://localhost:${HTTP_PORT}/get`,
+        }).then(body => {
+          expect(body).to.eq('It worked!')
+          if (this.debugProxy) {
+            expect(this.debugProxy.requests[0]).to.include({
+              url: `http://localhost:${HTTP_PORT}/get`
+            })
+          }
+        })
+      })
+
+      it('HTTPS pages can be loaded', function() {
+        return this.request({
+          url: `https://localhost:${HTTPS_PORT}/get`
+        }).then(body => {
+          expect(body).to.eq('It worked!')
+          if (this.debugProxy) {
+            expect(this.debugProxy.requests[0]).to.include({
+              https: true,
+              url: `localhost:${HTTPS_PORT}`
+            })
+          }
+        })
+      })
+
+      it('HTTP errors are catchable', function() {
+        return this.request({
+          url: `http://localhost:${HTTP_PORT}/empty-response`,
+        }).catch(err => {
+          if (this.debugProxy) {
+            expect(this.debugProxy.requests[0]).to.include({
+              url: `http://localhost:${HTTP_PORT}/empty-response`
+            })
+            expect(err.statusCode).to.eq(502)
+          } else {
+            expect(err.message).to.eq('Error: socket hang up')
+          }
+        })
+      })
+
+      it('HTTPS errors are catchable', function() {
+        return this.request({
+          url: `https://localhost:${HTTPS_PORT}/empty-response`,
+        }).catch(err => {
+          expect(err.message).to.eq('Error: socket hang up')
+        })
+      })
+
+      it('HTTP websocket connections can be established and used', function() {
+        return new Promise((resolve) => {
```

Comment:
due to the way mocha handles `unhandledRejections` and `uncaughtExceptions` this can create situations in which the errors don't immediately indicate what happened.

Because the `on(message, cb)` handler is called asynchronous - if you were to throw in here it would not be caught by the promise - it'll become a global uncaught exception event.

This can be mitigated by a number of ways like this...

```js
return new Promise((resolve) => {
  Io.client(`http://localhost:${HTTP_PORT}`, {
    agent: this.agent,
    transports: ['websocket']
  })
  .on('message', resolve)
})
.then((msg) => {
  expect(msg).to.eq('It worked!')
  
  if (this.debugProxy) {
    expect(this.debugProxy.requests[0].ws).to.be.true
    expect(this.debugProxy.requests[0].url).to.include('http://localhost:31080')
  }
})
```

Doing it this way will ensure that `msg` always resolves the promise which then allows you to carry that value forward and the promise chain is always hooked up correctly. I don't think we do this very often in other tests though - so you won't find very many good examples.